hibernate-core/src/test/java/org/hibernate/test/batchfetch/BatchFetchNotFoundIgnoreDefaultStyleTest.java

```diff
@@ -0,0 +1,339 @@
+		// Task with ID == 7 (because the batch is empty).
+		// query should have 1 parameter [7];
+		// Task with ID == 7 won't be found.
+		assertEquals( 1, paramterCounts.get( 3 ).intValue() );
+
+		assertEquals( NUMBER_OF_EMPLOYEES, employees.size() );
+		for ( int i = 0 ; i < NUMBER_OF_EMPLOYEES ; i++ ) {
+			if ( i == 1 || i == 7 ) {
+				assertNull( employees.get( i ).task );
+			}
+			else {
+				assertEquals( tasks.get( i ).id, employees.get( i ).task.id );
+			}
+		}
+	}
+
+	@Test
+	public void testMostNotFoundFromQuery() {
+
+		doInHibernate(
+				this::sessionFactory, session -> {
+					// delete all but last Task entity
+					for ( int i = 0; i < 7; i++ ) {
+						session.delete( tasks.get( i ) );
+					}
+				}
+		);
+
+		statementInspector.clear();
+
+		final List<Employee> employees = doInHibernate(
+				this::sessionFactory, session -> {
+					List<Employee> results =
+							session.createQuery( "from Employee e order by e.id", Employee.class ).getResultList();
+					for ( int i = 0 ; i < tasks.size() ; i++ ) {
+						checkInBatchFetchQueue( tasks.get( i ).id, session, false );
+					}
+					return results;
+				}
+		);
+
+		final List<Integer> paramterCounts = statementInspector.parameterCounts;
+
+		// there should be 8 SQL statements executed
+		assertEquals( 8, paramterCounts.size() );
+
+		// query loading Employee entities shouldn't have any parameters
+		assertEquals( 0, paramterCounts.get( 0 ).intValue() );
+
+		// query specifically for Task with ID == 0 will result in 1st batch;
+		// query should have 5 parameters for [0,1,2,3,4];
+		// Task with IDs == [0,1,2,3,4] won't be found
+		assertEquals( 5, paramterCounts.get( 1 ).intValue() );
+
+		// query specifically for Task with ID == 1 will result in 2nd batch;
+		// query should have 4 parameters [1,5,6,7];
+		// Task with IDs == [1,5,6] won't be found; Task with ID == 7 will be found.
+		assertEquals( 4, paramterCounts.get( 2 ).intValue() );
+
+		// query specifically for Task with ID == 2 will result in just querying
+		// Task with ID == 2 (because the batch is empty).
+		// query should have 1 parameter [2];
+		// Task with ID == 2 won't be found.
+		assertEquals( 1, paramterCounts.get( 3 ).intValue() );
+
+		// query specifically for Task with ID == 3 will result in just querying
+		// Task with ID == 3 (because the batch is empty).
+		// query should have 1 parameter [3];
+		// Task with ID == 3 won't be found.
+		assertEquals( 1, paramterCounts.get( 4 ).intValue() );
+
+		// query specifically for Task with ID == 4 will result in just querying
+		// Task with ID == 4 (because the batch is empty).
+		// query should have 1 parameter [4];
+		// Task with ID == 4 won't be found.
+		assertEquals( 1, paramterCounts.get( 5 ).intValue() );
+
+		// query specifically for Task with ID == 5 will result in just querying
+		// Task with ID == 5 (because the batch is empty).
+		// query should have 1 parameter [5];
+		// Task with ID == 5 won't be found.
```

Comment:
Without the fix, the following IDs would be queried in the batch: [5,4,6,2,3], causing the following to fail.

After the fix, the batch would be empty, so Hibernate would simply try to load the entity with id == 5 (since it was not found in the last batch). This would perform much better than before the fix.